File: problems/leetcode/1200/1235/1235.cpp
// 1235. Maximum Profit in Job Scheduling
// https://leetcode.com/problems/maximum-profit-in-job-scheduling/description/

using namespace std;

#include <map>
#include <vector>

class Solution {
public:
  int jobScheduling(vector<int> &startTime, vector<int> &endTime,
                    vector<int> &profit) {
    int n = startTime.size();

    vector<vector<int>> jobs;
    for (int i = 0; i < n; i++) {
      jobs.push_back({startTime[i], endTime[i], profit[i]});
    }
    sort(jobs.begin(), jobs.end(), greater<vector<int>>());

    int currentMaxProfit = jobs[0][2];
    map<int, int> maxProfitWithStartTime;
    maxProfitWithStartTime[jobs[0][0]] = currentMaxProfit;

    for (int i = 0; i < n; i++) {
      int et = jobs[i][1], st = jobs[i][0], p = jobs[i][2];

      map<int, int>::iterator iter = maxProfitWithStartTime.lower_bound(et);

      if (iter == maxProfitWithStartTime.end()) {
        currentMaxProfit = max(currentMaxProfit, p);
        maxProfitWithStartTime[st] =
            max(maxProfitWithStartTime[st], currentMaxProfit);
        continue;
      }

      currentMaxProfit = max(currentMaxProfit, p + iter->second);
      maxProfitWithStartTime[st] = currentMaxProfit;
    }

    return currentMaxProfit;
  }
};

// dp[n]을 시간 n에서 시작해서 얻을 수 있는 최대 profit이라고 하자.
// 어차피 startTime에 등장하는 시간의 종류가 정해져 있기 때문에, 큰 startTime을
// 가진 job부터 순회하면서 아래의 동작을 반복한다. job의 startTime, endTime,
// profit에 대해서 dp[startTime] = max(profit + dp[lower_bound(endTime)],
// ...(startTime <= t < endTime 에 존재하는 모든 dp[k]들)) 특정 job을 선택했을
// 때의 profit과, 시간에 따른 profit을 구분해야 한다.

File: problems/leetcode/1200/1235/1235.py
# 1235. Maximum Profit in Job Scheduling
# https://leetcode.com/problems/maximum-profit-in-job-scheduling/description/

from typing import *

# endTime의 오름차순으로 정렬
# 작은 순서대로 확인하면, 우선 매번 차례가 온 Job의 startTime 이전에 완료된 모든 Job들을 거쳐온 상태임
# Binary Search를 이용해서 이전의 최대값을 구하고, 반드시 누적 profit들이 오름차순이 되게 만들기.
# 동일한 endTime에 대한 profit이 이미 있고 더 높은 profit을 알게 되면 무조건 바꿔야 함.

class Solution:
    def jobScheduling(self, startTime: List[int], endTime: List[int], profit: List[int]) -> int:
        jobs = list(sorted([(endTime[i], startTime[i], profit[i]) for i in range(len(startTime))]))
        cumulativeProfits = [] # [endTime, profit]

        def findLargestProfit(et: int) -> int:
            start, end = 0, len(cumulativeProfits)
            target = et + 1

            while start < end:
                mid = (start + end) // 2
                if cumulativeProfits[mid][0] < target:
                    start = mid + 1
                else:
                    end = mid

            return cumulativeProfits[start - 1][1] if start > 0 else 0

        for e, s, p in jobs:
            maxProfit = p + findLargestProfit(s)

            if len(cumulativeProfits) > 0 and maxProfit <= cumulativeProfits[-1][1]:
                continue

            if len(cumulativeProfits) == 0 or cumulativeProfits[-1][0] < e:
                cumulativeProfits.append([e, maxProfit])
            else:
                cumulativeProfits[-1][1] = maxProfit

        return cumulativeProfits[-1][1]
